src/core/handle.ml

```diff
@@ -305,66 +336,125 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
-      let finalize ss st =
-        Console.State.pop ();
-        match pe.elt with
-        | P_proof_abort ->
-            (* Just ignore the command, with a warning. *)
-            wrn cmd.pos "Proof aborted."; ss
-        | P_proof_admit ->
-            (* If the proof is finished, display a warning. *)
-            if Proof.finished st then
-              wrn cmd.pos "The proof is finished. You can use 'qed' instead.";
-            (* Add a symbol corresponding to the proof, with a warning. *)
-            out 3 "(symb) %s (admit)\n" x.elt;
-            wrn cmd.pos "Proof admitted.";
-            fst (Sig_state.add_symbol ss expo Const Eager x a impl None)
-        | P_proof_qed   ->
-            (* Check that the proof is indeed finished. *)
-            if not (Proof.finished st) then
-              begin
-                let _ = Tactics.handle_tactic ss st (none P_tac_print) in
-                fatal cmd.pos "The proof is not finished."
-              end;
-            (* Add a symbol corresponding to the proof. *)
-            out 3 "(symb) %s (qed)\n" x.elt;
-            fst (Sig_state.add_symbol ss expo Const Eager x a impl None)
+          match ao with
+          | Some ao ->
+            let a = scope_basic expo ao in
+            let metas_a = Basics.get_metas true a in
+            Some a, metas_a
+          | None -> None, MetaSet.empty
+        end
       in
-      let data =
-        { pdata_stmt_pos = stmt.pos ; pdata_p_state = st ; pdata_tactics = ts
-        ; pdata_finalize = finalize ; pdata_term_pos = pe.pos }
+      (* If a type [ao = Some a] is given, then we check that it is
+         typable by a sort and that [t] has type [a]. Otherwise, we
+             try to infer thetype of [t]. Unification goals are collected *)
+      (* Proof script *)
+      let (ts,pe) =
+        let p_end = Pos.make cmd.pos P_proof_end in
+        let _solve = Pos.make cmd.pos P_unif_solve in
+        begin
+          match p_t,ts_pe with
+          | Some p_t,None ->
+            [Pos.make cmd.pos (P_tac_refine(p_t))], p_end
+          | Some p_t,Some(ts,pe) ->
+            Pos.make cmd.pos (P_tac_refine(p_t))::ts, pe
+          | None,Some(ts,pe) -> ts,pe
+          | None,None -> [],p_end
+        end
       in
-      (ss, Some(data))
+      let metas_a_t = MetaSet.union metas_a MetaSet.empty in
+      let add_goal m =
+        List.insert Proof.Goal.compare (Proof.Goal.goal_typ_of_meta m)
+      in
+      let typ_goals_a = MetaSet.fold add_goal metas_a_t [] in
+      let typ_goals_a = List.map Proof.Goal.typ typ_goals_a in
+      let proof_term,goals,sig_symbol,pdata_expo =
+        let sort_goals, a = Proof.goals_of_typ x.pos ao t in
+        (* And the main "type" goal *)
+        let proof_term,typ_goal =
+          match e with
+          | Def ->
+            let proof_term = Meta.fresh ~name:x.elt a 0 in
+            let proof_goal = Proof.goal_of_meta proof_term in
+            Some proof_term,[proof_goal] @ typ_goals_a
```

Comment:
can be simplified to: ` proof_goal :: typ_goals_a`

Reply:
fixed